iavl/iavl_proof2.go

```diff
@@ -0,0 +1,251 @@
+	if proof.LeftPath == nil {
+		if !proof.RightPath.IsLeftmost() {
+			return errors.New("right path is only one but not leftmost")
+		}
+	}
+
+	// Only left path exists, check that node is at right boundary.
+	if proof.RightPath == nil {
+		if !proof.LeftPath.IsRightmost() {
+			return errors.New("left path is only one but not rightmost")
+		}
+	}
+
+	return nil
+}
+
+type KeyRangeExistsProof struct {
+	RootHash   []byte
+	PathToKeys []*PathToKey
+}
+
+func (proof *KeyRangeExistsProof) Verify(key []byte, value []byte, root []byte) bool {
+	return false
+}
+
+func ReadKeyExistsProof(data []byte) (*KeyExistsProof, error) {
+	proof := new(KeyExistsProof)
+	err := wire.ReadBinaryBytes(data, &proof)
+	if err != nil {
+		return nil, err
+	}
+	return proof, nil
+}
+
+func (node *IAVLNode) constructKeyExistsProof(t *IAVLTree, key []byte, proof *KeyExistsProof) (value []byte, exists bool) {
+	if node.height == 0 {
+		if bytes.Compare(node.key, key) == 0 {
+			proof.LeafHash = node.hash
+			return node.value, true
+		}
+		return nil, false
+	}
+
+	if bytes.Compare(key, node.key) < 0 {
+		if value, exists := node.getLeftNode(t).constructKeyExistsProof(t, key, proof); exists {
+			branch := IAVLProofInnerNode{
+				Height: node.height,
+				Size:   node.size,
+				Left:   nil,
+				Right:  node.getRightNode(t).hash,
+			}
+			proof.InnerNodes = append(proof.InnerNodes, branch)
+			return value, true
+		}
+		return nil, false
+	}
+
+	if value, exists = node.getRightNode(t).constructKeyExistsProof(t, key, proof); exists {
+		branch := IAVLProofInnerNode{
+			Height: node.height,
+			Size:   node.size,
+			Left:   node.getLeftNode(t).hash,
+			Right:  nil,
+		}
+		proof.InnerNodes = append(proof.InnerNodes, branch)
+		return value, true
+	}
+	return nil, false
+}
+
+func (node *IAVLNode) constructKeyNotExistsProof(t *IAVLTree, key []byte, proof *KeyNotExistsProof) error {
+	idx, _, exists := t.Get(key)
+	if exists {
+		return errors.Errorf("couldn't construct non-existence proof: key 0x%x exists", key)
+	}
+
+	var (
+		lkey, lval []byte
+		rkey, rval []byte
+	)
+	if idx > 0 {
+		lkey, lval = t.GetByIndex(idx - 1)
+	}
+	if idx < t.Size()-1 {
+		rkey, rval = t.GetByIndex(idx)
+	}
+
+	// TODO: What if tree is empty?
+	if lkey == nil && rkey == nil {
+		return errors.New("couldn't get keys required for non-existence proof")
+	}
+
+	if lkey != nil {
+		lproof := &KeyExistsProof{
+			RootHash: t.root.hash,
+		}
+		node.constructKeyExistsProof(t, lkey, lproof)
+		proof.LeftPath = &lproof.PathToKey
+		proof.LeftNode = IAVLProofLeafNode{KeyBytes: lkey, ValueBytes: lval}
+	}
+	if rkey != nil {
+		rproof := &KeyExistsProof{
+			RootHash: t.root.hash,
+		}
+		node.constructKeyExistsProof(t, rkey, rproof)
+		proof.RightPath = &rproof.PathToKey
+		proof.RightNode = IAVLProofLeafNode{KeyBytes: rkey, ValueBytes: rval}
+	}
+
+	return nil
+}
+
+func (t *IAVLTree) ConstructKeyExistsProof(key []byte) (value []byte, proof *KeyExistsProof) {
```

Comment:
Do these IAVLTree.Construct* methods need to be public?  Shall we only expose them through `GetWithProof()`... or whatever you call the method?

I like to keep the public API relatively minimal, like only one method to do a given action.